### PR TITLE
edited regex to handle whitespace,added tests

### PR DESCRIPTION
**Pull request by arinkulshi-skylight**
# BACKEND PULL REQUEST

## Related Issue
https://app.zenhub.com/workspaces/simplereport-2025-and-onwards-605b43a6ceb92c000f86279a/issues/gh/cdcgov/prime-simplereport/8253

## Changes Proposed

extra whitespace (e.g. "10/18/2024 10:00" - passes but with extra whitespaces it fails)

## Additional Information
This ticket does not handle AM/PM. After researching more into this I believe this change would be more then 1-2 points. 

## Testing

- How should reviewers verify this PR?
https://github.com/CDCgov/prime-simplereport/pull/8225#issuecomment-2445401198 please see the steps here to reproduce as well as the csv.

### CODE REVIEW

File: backend/src/main/java/gov/cdc/usds/simplereport/validators/CsvValidatorUtils.java [hunk 1]
```diff
@@ -98,7 +98,7 @@ public class CsvValidatorUtils {
    * @see gov.cdc.usds.simplereport.utils.DateTimeUtils
    */
   private static final String DATE_TIME_REGEX =
-      "^(0{0,1}[1-9]|1[0-2])\\/(0{0,1}[1-9]|1\\d|2\\d|3[01])\\/\\d{4}( ([0-1]?[0-9]|2[0-3]):[0-5][0-9]( \\S+)?)?$";
+      "^(0{0,1}[1-9]|1[0-2])\\s*/\\s*(0{0,1}[1-9]|1\\d|2\\d|3[01])\\s*/\\s*\\d{4}(\\s+([0-1]?[0-9]|2[0-3])\\s*:\\s*[0-5][0-9](\\s+\\S+)?)?\\s*$";
```
Comment (arinkulshi-skylight) [inline backend/src/main/java/gov/cdc/usds/simplereport/validators/CsvValidatorUtils.java:101]: https://regex101.com/r/NwJqzI/1 you can test out the regex here!

File: backend/src/main/java/gov/cdc/usds/simplereport/utils/DateTimeUtils.java [hunk 2]
```diff
@@ -24,8 +24,18 @@ public class DateTimeUtils {
...
+
   public static final String TIMEZONE_SUFFIX_REGEX =
-      "^(0?[1-9]|1[0-2])/(0?[1-9]|1\\d|2\\d|3[01])/\\d{4}( ([0-1]?\\d|2[0-3]):[0-5]\\d)( \\S+)$";
+      "^"
+          + MONTH
+          + "\\s*/\\s*"
+          + DAY
+          + "\\s*/\\s*\\d{4}\\s+"
+          + HOUR
+          + "\\s*:\\s*[0-5]\\d\\s+(\\S+)\\s*$";
```
Comment (mpbrown) [inline backend/src/main/java/gov/cdc/usds/simplereport/utils/DateTimeUtils.java:38]: Nice way to break up this regex to make it more readable! Any particular reason for doing the same to year, minutes, and timezone code?

File: backend/src/main/java/gov/cdc/usds/simplereport/utils/DateTimeUtils.java [hunk 3]
```diff
@@ -147,8 +160,13 @@ public static ZoneId parseZoneId(String timezoneCode) {
   public static LocalDateTime parseLocalDateTime(
       String value, DateTimeFormatter dateTimeFormatter) {
     String dateTimeString = value;
+    dateTimeString =
+        dateTimeString.replace(" / ", "/").replace(" :", ":").replace(": ", ":").trim();
```
Comment (mpbrown) [inline backend/src/main/java/gov/cdc/usds/simplereport/utils/DateTimeUtils.java:164]: Should this also handle `"/ "` and `" /"`?

**Review by mpbrown — DISMISSED** [review]
Looks great overall and I appreciate the additional context you added in the description! Left some non-blocking questions

File: backend/src/main/java/gov/cdc/usds/simplereport/utils/DateTimeUtils.java (hunk 2 above)
Comment (arinkulshi-skylight) [reply backend/src/main/java/gov/cdc/usds/simplereport/utils/DateTimeUtils.java]: Good point! I have made that change.

File: backend/src/main/java/gov/cdc/usds/simplereport/utils/DateTimeUtils.java (hunk 3 above)
Comment (arinkulshi-skylight) [reply backend/src/main/java/gov/cdc/usds/simplereport/utils/DateTimeUtils.java]: I think it is ok to add those extra explicit checks just in case. I would assume they have a minimal impact on performance.  I have added all possible combinations to` .replace` including trailing whitespace. Let me know if I missed anything!

File: backend/src/main/java/gov/cdc/usds/simplereport/validators/CsvValidatorUtils.java (hunk 1 above)
Comment (mehansen) [reply backend/src/main/java/gov/cdc/usds/simplereport/validators/CsvValidatorUtils.java:101]: should `DATE_TIME_REGEX` and `TIMEZONE_SUFFIX_REGEX` essentially be the same except that the time zone is optional in the first one?

File: backend/src/main/java/gov/cdc/usds/simplereport/utils/DateTimeUtils.java [hunk 4]
```diff
@@ -121,10 +140,13 @@ public static ZonedDateTime convertToZonedDateTime(String dateString) {
...
-  private static ZoneId parseDateStringZoneId(String dateString) {
+  public static ZoneId parseDateStringZoneId(String dateString) {
     ZoneId zoneId;
-
-    var timezoneCode = dateString.substring(dateString.lastIndexOf(" ")).trim();
+    int lastSpaceIndex = dateString.lastIndexOf(" ");
+    if (lastSpaceIndex < 0) {
+      return FALLBACK_TIMEZONE_ID;
+    }
+    var timezoneCode = dateString.substring(lastSpaceIndex + 1).trim();
```
Comment (mehansen) [inline backend/src/main/java/gov/cdc/usds/simplereport/utils/DateTimeUtils.java:149]: I didn't see this as a test case but we might want to handle the case where there is trailing whitespace as well - I think right now that would result in just using the default time zone

File: backend/src/main/java/gov/cdc/usds/simplereport/utils/DateTimeUtils.java [hunk 5]
```diff
@@ -147,8 +169,22 @@ public static ZoneId parseZoneId(String timezoneCode) {
...
+    dateTimeString =
+        dateTimeString
+            .replace(" /", "/")
+            .replace("/ ", "/")
+            .replace("  /", "/") // Handles ("/" with multiple spaces before)
+            .replace("/  ", "/") // Handles ("/" with multiple spaces after)
+            .replace(" :", ":") // Removes space before ":"
+            .replace(": ", ":") // Removes space after ":"
+            .replace("  :", ":") // Handles (":" with multiple spaces before)
+            .replace(":  ", ":") // Handles (":" with multiple spaces after)
```
Comment (mehansen) [inline backend/src/main/java/gov/cdc/usds/simplereport/utils/DateTimeUtils.java:181]: is there a number of spaces that we are accepting vs not? 😅  
`01/01/    2023 11:11 PT` will pass our validator regex but then I don't think this would parse that properly

**Review by mehansen — COMMENTED** [review]
thanks! left a couple comments that maybe we want product input on? in terms of what's valid input

also: do we want to handle the case where there is preceding whitespace?

File: backend/src/main/java/gov/cdc/usds/simplereport/validators/CsvValidatorUtils.java (hunk 1 above)
Comment (arinkulshi-skylight) [reply backend/src/main/java/gov/cdc/usds/simplereport/validators/CsvValidatorUtils.java:101]: Initially this also made sense to me. After looking into this I think there are a couple reasons why they have been separated out. Timezone validation has an extra layer (checking valid timezone codes etc). We also seem to validate datetime with and without timezone by invoking a separate set of methods in each case. When I tried to combine the two patterns I was struggling to handle some of the wierd edge cases. In one of the test cases I got this: 

```
Test: testConvertToZonedDateTime_withFallback
Expected: 2023-06-28T14:00-04:00[US/Eastern]
Actual: 2023-06-28T12:00-04:00[US/Eastern]
```

When I tried to combine the two. It made it difficult to handle timezone conversions correctly. The system will think the parsing of the time is the timezone. When we handle it separately it allows us to process the two components indep of each other. I'll def check with the team on this but my feeling is it might be ok to keep it separate for now.

File: backend/src/main/java/gov/cdc/usds/simplereport/utils/DateTimeUtils.java (hunk 5 above)
Comment (arinkulshi-skylight) [reply backend/src/main/java/gov/cdc/usds/simplereport/utils/DateTimeUtils.java:181]: That is a good point I tried to cover as many of the scenarios as I could think of. I'll ask the team about this.

Comment (mehansen) [reply backend/src/main/java/gov/cdc/usds/simplereport/utils/DateTimeUtils.java:181]: sounds good! 👍 Imo whatever passes the validation step (matches the validation regex) should be able to be parsed since we would want to avoid exceptions on the parse step since they don't get returned back neatly to the user (or sometimes not shown to them at all)

File: backend/src/main/java/gov/cdc/usds/simplereport/utils/DateTimeUtils.java (hunk 4 above)
Comment (mehansen) [reply backend/src/main/java/gov/cdc/usds/simplereport/utils/DateTimeUtils.java:149]: I meant to say that while it's not explicitly called out in the ticket I think we should handle the case where we parse time zone properly even if there is trailing whitespace. basically, I think this example test should pass but it is failing for me locally:

```
  void exampleTest() {
    String dateString = "01/01/2023 11:11 PT ";
    ZoneId expectedZoneId = ZoneId.of("US/Pacific");
    ZoneId actualZoneId = parseDateStringZoneId(dateString);
    assertThat(actualZoneId).isEqualTo(expectedZoneId);
  }
```

File: backend/src/main/java/gov/cdc/usds/simplereport/validators/CsvValidatorUtils.java (hunk 1 above)
Comment (mehansen) [reply backend/src/main/java/gov/cdc/usds/simplereport/validators/CsvValidatorUtils.java:101]: ah sorry I didn't mean that they could be consolidated into a single expression - more that we could probably update the regex definitions to share components something like below (disclaimer: did not test this 😅 ) or using the `MONTH`, `DAY`, `HOUR`, etc variables that you defined

```
DATE_REGEX = "((0?[1-9]|1[0-2])\\s*/\\s*(0{0,1}[1-9]|1\\d|2\\d|3[01])\\s*/\\s*\\d{4}";
TIME_REGEX = "([0-1]?[0-9]|2[0-3])\\s*:\\s*[0-5][0-9]";
TIMEZONE_REGEX = "(\\S+)";
DATE_TIME_REGEX = "^" + DATE_REGEX + "\\s+(" + TIME_REGEX + "\\s+(" + TIMEZONE_REGEX + ")?)?\\s*$";
TIMEZONE_SUFFIX_REGEX = "^" + DATE_REGEX + "\\s+" + TIME_REGEX + "\\s+" + TIMEZONE_REGEX + "\\s*$";
```

that way in the future if we update one we're more likely to update the other one correctly as well

**Review by bobbywells52 — APPROVED** [review]
Tested and working as expected! Thanks for your great work on this... Regex + date time is a starter ticket from hell lol

File: backend/src/main/java/gov/cdc/usds/simplereport/utils/DateTimeUtils.java (hunk 4 above)
Comment (arinkulshi-skylight) [reply backend/src/main/java/gov/cdc/usds/simplereport/utils/DateTimeUtils.java]: That is a good catch. Apologies for missing it the first time.  I have edited the logic to address this and added test for these scenarios: 
        ```
    "01/01/2023 11:11 PT ", ZoneId.of("US/Pacific"),
            "01/01/2023 11:11 EST  ", ZoneId.of("US/Eastern"),
            "01/01/2023 11:11 MST\t", ZoneId.of("US/Mountain"),
            "01/01/2023 11:11 CST \t ", ZoneId.of("US/Central"),
            "01/01/2023 11:11 HST    ", ZoneId.of("US/Hawaii"));
```

**Comment by arinkulshi-skylight** [comment]
We have decided to close this PR for the moment until we get the final input from Mindy. https://skylight-hq.slack.com/archives/C065V4DHBQF/p1739215582181589. We will revisit this change after her approval. Thanks for all the great comments and help on this ticket!
